lib/SILOptimizer/Mandatory/DiagnoseInfiniteRecursion.cpp

```diff
@@ -0,0 +1,242 @@
+            SILWitnessTable *WT;
+
+            std::tie(F, WT) = M.lookUpFunctionInWitnessTable(
+                WMI->getConformance(), WMI->getMember());
+            if (F == Target)
+              return true;
+
+            continue;
+          }
+        }
+      }
+      return false;
+    }
+
+    using ExitBlockSet = llvm::SmallPtrSetImpl<SILBasicBlock *>;
+    static bool hasInfinitelyRecursiveApply(SILFunction &Fn,
+                                            SILFunction *TargetFn,
+                                            const ExitBlockSet &ExitBlocks) {
+      enum RecursiveState {
+        FoundRecursion          = 0,
+        FoundPathOutOfFunction  = 1,
+      };
+
+      // Set the entry block into the state map.
+      llvm::DenseMap<SILBasicBlock *, RecursiveState> States;
+      States[Fn.getEntryBlock()] = FoundPathOutOfFunction;
+
+      // Make the processing stack and seed it with the entry block.
+      SmallVector<SILBasicBlock *, 16> Stack;
+      Stack.push_back(Fn.getEntryBlock());
+
+      while (!Stack.empty()) {
+        SILBasicBlock *CurBlock = Stack.pop_back_val();
+        RecursiveState &CurState = States[CurBlock];
+
+        if (CurState == FoundPathOutOfFunction) {
+          // Only change state if the block has a recursive call.
+          if (hasRecursiveCallInPath(*CurBlock, TargetFn)) {
+            CurState = FoundRecursion;
+          } else if (ExitBlocks.count(CurBlock)) {
+            // Found a path to the exit node without a recursive call.
+            return false;
+          }
+        }
+
+        // Loop over successor blocks and add them to the Stack if their state
+        // changes.
+        auto Iter = CurBlock->succ_begin();
+        for (auto EndIter = CurBlock->succ_end(); Iter != EndIter; ++Iter) {
+          if (SILBasicBlock *Succ = *Iter) {
+            if (States[Succ] < CurState) {
+              States[Succ] = CurState;
+              Stack.push_back(Succ);
+            }
+          }
+        }
+      }
+
+      // Bail out if all exit nodes are reachable, and only reachable, through
+      // a recursive call.
+      for (auto *EB : ExitBlocks) {
+        if (States[EB] == FoundPathOutOfFunction)
+          return false;
+      }
+      return true;
+    }
+
+    static SILFunction *computeLazyGetterClosure(PostOrderFunctionInfo *PO) {
+      for (auto *BB : PO->getReversePostOrder()) {
+        for (auto &I : *BB) {
+          auto *AI = dyn_cast<ApplyInst>(&I);
+          if (!AI) {
+            continue;
+          }
+
+          auto *Callee = AI->getCalleeFunction();
+          if (!Callee) {
+            continue;
+          }
+
+          // Ignore external declarations.
+          if (Callee->empty() || Callee->isExternalDeclaration()) {
+            continue;
+          }
+
+          // Ignore non-closure callees.
+          if (!Callee->getLocation().isASTNode<AbstractClosureExpr>()) {
+            continue;
+          }
+
+          return Callee;
```

Comment:
It looks like you're simply returning the first function call based on its debug location. Using the debug location is *probably* ok in a diagnostic pass, but I'm not 100% sure. Can you comment on why simply looking for the first function call is sound or conservative? Are you effectively disabling your diagnostics on some class of closures in general?

Reply:
The goal is just to weed out the apply part of this

```swift
        // class Klass {
        //   lazy var recur: String = {
        //     // ...
        //     print(recur)
        //     // ...
        //   }()
        // }
```

I admit the approach is hamfisted because I wasn't able to find convincing alternative.  I'm open to suggestions here.